lib/metrics/index.js

```diff
@@ -0,0 +1,144 @@
+'use strict';
+
+const PrometheusClient = require('./clients/prometheus');
+const StatsDClient = require('./clients/statsd');
+const LogClient = require('./clients/log');
+const Metric = require('./metric');
+const normalizeName = require('./utils').normalizeName;
+
+const SUPPORTED_CLIENTS = {
+    PROMETHEUS: 'prometheus',
+    STATSD: 'statsd',
+    LOG: 'log'
+};
+
+const DEPRECATED_METHODS = ['timing', 'increment', 'decrement', 'gauge', 'unique'];
+
+class Metrics {
+    constructor(options, logger) {
+        this.logger = logger;
+        this.cache = new Map();
+        this.clients = [];
+        options.forEach((o) => {
+            if (o.type === SUPPORTED_CLIENTS.PROMETHEUS) {
+                this.clients.push(new PrometheusClient(o, this.logger));
+            }
+            if (o.type === SUPPORTED_CLIENTS.STATSD) {
+                this.clients.push(new StatsDClient(o, this.logger));
+            }
+            if (o.type === SUPPORTED_CLIENTS.LOG) {
+                o.methods = DEPRECATED_METHODS;
+                this.clients.push(new LogClient(o, this.logger));
+            }
+        });
+    }
+
+    fetchClient(name) {
+        return this.clients.find((client) => {
+            return client.constructor.name === name;
+        });
+    }
+
+// Example Options:
+// {
+//     type: 'Counter',
+//     name: 'hitcount',
+//     prometheus: {
+//         name: 'hitcount',
+//         help: 'hit count',
+//         buckets: [], // https://github.com/siimon/prom-client#histogram
+//         percentiles: [], // https://github.com/siimon/prom-client#summary
+//     },
+//     labels: {
+//         keys: [],
+//         labelPosition: 'before',
+//         omitLabelNames: false
+//     }
+// }
+    makeMetric(options) {
+        const metric = new Metric(this.clients, this.logger, options);
+        this.cache.set(options.name, metric);
+        return metric;
+    }
+
+    fetchMetric(name) {
+        return this.cache.get(name);
+    }
```

Comment:
Why don't we merge the `makeMetric` and `fetchMetric` together, and lazy-create it on demand? What's the point of having 2 sedate methods?

Reply:
makeMetric() wants the full object in order to initialize all configured metrics clients.  I figured that providing this object every time you wanted to get one the user already initialized would be onerous.  fetchMetric() is a getter into the cache so that users need not implement their own cache.

Reply:
I think too much focus in this PR is given to Prometheus. We should assume sensible defaults and use them. Yes, Prometheus is powerful, but personally I'd want to see more focus on client simplicity and compatibility. If one does need all Prometheus has to offer, it's fine to have a class here that can expose all of that, and then the client can use it and tell service-runner "use this object for metrics", since I figure that 99.99% of clients is perfectly fine with "send metric X with value Y".